### ISSUE TITLE
Convert Python Object in Javascript/JSON

### ISSUE THREAD

**Issue by TA2k** [extremeheat/JSPyBridge#60]
I'm struggling at the moment to convert python object in a javascript usable object or json

With json dumps I have no idea how I can realize the custom BytesEncoder in Javascript
```
class BytesEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, bytes):
            return obj.decode('utf-8')
        return json.JSONEncoder.default(self, obj)

json.dumps(states.__dict__, cls=BytesEncoder)
```
The option via dir(states) is not working because dir is not available and not working via eval

```
    const states = {};
    const keys = await py`dir(${states})`;
    for await (let [index, key] of await py.enumerate(keys)) {
         if (key.startsWith("__")) {
             continue;
         }
        state[key] = await appliance_state[key]
    }
```

Is there any example or recommendation?

**Comment by TheCoder76-Codes** [comment]
Hi @TA2k ,
I've also encountered your problem, and I have managed to find a solution.
**TL;DR:**
```js
let x = // your dict
x = JSON.parse(String(x).replaceAll("'", '"'))
```
**Long Explanation**
In my case, I was getting a dictionary from a variable. I assigned the dictionary to a variable, which I will call x in this case.
```js
let x = // ...
```
When using the `typeof` statement, we can see that it returns a `function` type. We can just convert this to a string using the `String` object.
```js
x = String(x)
```
Now we can retry converting it to an object.
```js
x = JSON.parse(x)
```
However, we get an error, as JSON likes to have double quotation marks `"` instead of single ones `'`, which is what the dictionary uses. We can counter this by replacing all single quotation marks in the dictionary with double ones using the `replaceAll` method.

Altogether, it looks a bit like this:
```js
let x = // your dict
x = JSON.parse(String(x).replaceAll("'", '"'))
```
I hope this helps!

**Comment by TA2k** [comment]
It is a bit complex here my solution
```
  pythonToJson(objectString) {
    if (!objectString.replace) {
      this.log.warn("pythonToJson: objectString is not a string");
      return objectString;
    }
    objectString = objectString
      .replace(/b'[^']*'/g, "''")
      .replace(/: <[^<]*>,/g, ":'',")
      .replace(/{'_/g, `{'`)
      .replace(/, '_/g, `, '`)
      .replace(/'/g, `"`)
      .replace(/ None,/g, `null,`)
      .replace(/ True,/g, `true,`)
      .replace(/ False,/g, `false,`);

    this.log.debug(objectString);
    return objectString;
  }
  ```

**Comment by dkerr64** [comment]
> It is a bit complex here my solution
> 
> ```
>   pythonToJson(objectString) {
>     if (!objectString.replace) {
>       this.log.warn("pythonToJson: objectString is not a string");
>       return objectString;
>     }
>     objectString = objectString
>       .replace(/b'[^']*'/g, "''")
>       .replace(/: <[^<]*>,/g, ":'',")
>       .replace(/{'_/g, `{'`)
>       .replace(/, '_/g, `, '`)
>       .replace(/'/g, `"`)
>       .replace(/ None,/g, `null,`)
>       .replace(/ True,/g, `true,`)
>       .replace(/ False,/g, `false,`);
> 
>     this.log.debug(objectString);
>     return objectString;
>   }
> ```

I ran into a problem with this because it assumed that a comma appears after every `None`, `True` or `False`.  I had an object that ended with `{....: False}` so there was no comma, just the closing brace. So I replaced the tests with `.replaceAll(': False', ': false')` (ie, test for the colon/space before the value rather than the comma after the value.
